### PR TITLE
Refactor generic IPMI modules for reduced complexity

### PR DESCRIPTION
**Pull request by cfernald**
- Remove duplicate code in DXE/PEI implementations.
- Remove unused definitions.
- Unified interface definitions to improve code re-use.
- Moved to PCDs over static definitions for timeouts/retries.
- Removed redundant components.

### CODE REVIEW

File: IpmiFeaturePkg/GenericIpmi/Dxe/DxeGenericIpmi.c [hunk 1]
```diff
@@ -1,30 +1,351 @@
...
+  }
+
+  return Status;
+}
+
+/**
  @brief
-  This is entry point for IPMI service for BIOS POST.
+  This is entry point for IPMI service for DXE. Initializes the BMC information
+  and prepares transport and protocol for use. Depending on the configuration,
```
Comment (os-d) [inline IpmiFeaturePkg/GenericIpmi/Dxe/DxeGenericIpmi.c:347]: Looks like this comment got cutoff?

**Review by os-d — APPROVED** [review]

File: IpmiFeaturePkg/GenericIpmi/Common/GenericIpmiCommon.h [hunk 2]
```diff
@@ -9,7 +9,7 @@
 #ifndef _IPMI_COMMON_BMC_H_
 #define _IPMI_COMMON_BMC_H_
 
-#define MAX_TEMP_DATA      255// 160 Modified to increase number of bytes transfered per command
+#define MAX_TEMP_DATA      255
 #define BMC_SLAVE_ADDRESS  0x20
 #define MAX_SOFT_COUNT     10
 #define COMP_CODE_NORMAL   0x00
```
[the remark this answers is not in the document]
Comment (cfernald) [reply IpmiFeaturePkg/GenericIpmi/Common/GenericIpmiCommon.h:15]: I'm actually going to remove all of this, becuase it duplicate with the basecore ipmi.h

File: IpmiFeaturePkg/Library/BmcSelfTestPlatformLibNull/BmcSelfTestPlatformLibNull.c [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  @param[out] BmcStatus   A pointer that receives the BMC status change if needed.
+
+  @retval   EFI_SUCCESS                 <Description>
+**/
+VOID
+GetDeviceSpecificTestResults (
+  IN  IPMI_SELF_TEST_RESULT_RESPONSE  *Result,
+  OUT BMC_STATUS                      *BmcStatus
+  )
+{
```
[the remark this answers is not in the document]
Comment (cfernald) [reply IpmiFeaturePkg/Library/BmcSelfTestPlatformLibNull/BmcSelfTestPlatformLibNull.c:26]: Good point, i think its probably best not to ignore un-known reponses here.

File: IpmiFeaturePkg/GenericIpmi/Pei/PeiGenericIpmi.c [hunk 4]
```diff
@@ -61,15 +64,15 @@ PeiInitializeIpmiPhysicalLayer (
...
 
-  mIpmiInstance->PeiIpmiBmcDataDesc.Flags = EFI_PEI_PPI_DESCRIPTOR_PPI | EFI_PEI_PPI_DESCRIPTOR_TERMINATE_LIST;
-  mIpmiInstance->PeiIpmiBmcDataDesc.Guid  = &gPeiIpmiTransportPpiGuid;
-  mIpmiInstance->PeiIpmiBmcDataDesc.Ppi   = &mIpmiInstance->IpmiTransportPpi;
+  gPeiIpmiBmcDataDesc.Flags = EFI_PEI_PPI_DESCRIPTOR_PPI | EFI_PEI_PPI_DESCRIPTOR_TERMINATE_LIST;
+  gPeiIpmiBmcDataDesc.Guid  = &gPeiIpmiTransportPpiGuid;
+  gPeiIpmiBmcDataDesc.Ppi   = &mIpmiInstance->IpmiTransport;
 
   //
   // Initialize the transport layer.
```
[the remark this answers is not in the document]
Comment (cfernald) [reply IpmiFeaturePkg/GenericIpmi/Pei/PeiGenericIpmi.c:78]: Good catch

File: IpmiFeaturePkg/GenericIpmi/Dxe/DxeGenericIpmi.c [hunk 5]
```diff
@@ -1,51 +1,447 @@
...
+  //
+  if (EFI_ERROR (Status) || (Retries == 0)) {
+    DEBUG ((DEBUG_ERROR, "\n[IPMI]  BMC self-test does not respond (status: %r)!\n\n", Status));
+    if (*ErrorCount < MAX_SOFT_COUNT) {
+      StatusCodeValue[*ErrorCount] = EFI_COMPUTING_UNIT_FIRMWARE_PROCESSOR | EFI_CU_FP_EC_COMM_ERROR;
+      (*ErrorCount)++;
+    }
+
+    IpmiInstance->BmcStatus = BMC_HARDFAIL;
+    return Status;
```
Comment (kuqin12) [inline IpmiFeaturePkg/GenericIpmi/Dxe/DxeGenericIpmi.c:132]: Probably want to set it to timeout error if the retry count expires?

File: IpmiFeaturePkg/GenericIpmi/Dxe/DxeGenericIpmi.c [hunk 6]
```diff
@@ -1,51 +1,447 @@
...
+    if ((pBmcExecContext->CurrentExecutionContext == IPMI_BMC_IN_FORCED_UPDATE_MODE) &&
+        !EFI_ERROR (Status))
+    {
+      DEBUG ((DEBUG_WARN, "[IPMI] BMC in Forced Update mode, skip waiting for BMC_READY.\n"));
+      IpmiInstance->BmcStatus = BMC_UPDATE_IN_PROGRESS;
+    } else {
+      //
+      // Updatemode = 1 mean BMC is not ready, continue waiting.
+      //
+      while (Retries-- != 0) {
```
Comment (kuqin12) [inline IpmiFeaturePkg/GenericIpmi/Dxe/DxeGenericIpmi.c:313]: layman question: do you need to reset this counter for this operation?

File: IpmiFeaturePkg/GenericIpmi/Dxe/DxeGenericIpmi.c (hunk 5 above)
Comment (cfernald) [reply IpmiFeaturePkg/GenericIpmi/Dxe/DxeGenericIpmi.c:132]: I actually think its better to bubble up the underlying error code from then an error indicating we ran out of retries personally.

File: IpmiFeaturePkg/GenericIpmi/Dxe/DxeGenericIpmi.c (hunk 6 above)
Comment (cfernald) [reply IpmiFeaturePkg/GenericIpmi/Dxe/DxeGenericIpmi.c:313]: Yeah probably. will do

**Pull request merged**